File: 690-employee-importance/690-employee-importance.py
"""
# Definition for Employee.
class Employee:
    def __init__(self, id: int, importance: int, subordinates: List[int]):
        self.id = id
        self.importance = importance
        self.subordinates = subordinates
"""

class Solution:
    def getImportance(self, employees: List['Employee'], id: int) -> int:
        total = 0
        for emp in employees:
            if emp.id == id:
                for sub in emp.subordinates:
                    total += self.getImportance(employees,sub)
                return total + emp.importance

File: 690-employee-importance/690-employee-importance.cpp
/*
// Definition for Employee.
class Employee {
public:
    int id;
    int importance;
    vector<int> subordinates;
};
*/

class Solution {
public:
    int getImportance(vector<Employee*> employees, int id) {
        map<int, pair<int, vector<int>>> mp;
        for (auto e: employees) {
            mp[e->id].first = e->importance;
            for (auto x: e->subordinates) {
                mp[e->id].second.push_back(x);
            }
        }
        set<int>st;
        int cnt = 0;
        function<int(int)> dfs = [&](int root) {
            if (st.count(root))
                return 0;
            cnt += mp[root].first;
            for (auto x: mp[root].second) {
                dfs(x);
            }
            return cnt;
        };
        return dfs(id);
    }
};